asdf/_tests/_regtests/test_1520.py

```diff
@@ -0,0 +1,34 @@
+import numpy as np
+
+import asdf
+
+
+def test_failed_update_corrupts_file(tmp_path):
```

Comment:
Yes and no :)

The original issue was discovered while testing update and compressed data. If during an update a compressed block outgrew it's previous size the update would fail and the original file would be left in a 'corrupt' state where the tree did not match the blocks.

Ending up corrupting the original file always seems like a risk when calling 'update' (what would happen if halfway through writing the tree your computer lost power?).

If I recall correctly, I added the regtest prior to a working 'update' with the new block management code. While working on update the conditions that triggered the failed update were fixed (with the code in this PR compressed blocks can grow during an update and no error occurs). I ended up leaving in the test and modifying it to reflect that the update no longer fails.

I removed `test_1520` (but I think we should update and close the issue with this PR since the code that previously failed is fixed by this PR). I also added a new test `test_update_compressed_blocks` with some notes about the above history and to test that updating compressed data (causing it to increase size on disk) no longer causes update to fail: 590c9a35fa2bc7688801c297bd00ab59eb17117f